nin/backend/generate/generate.js

```diff
@@ -1,21 +1,66 @@
 var fs = require('fs'),
     path = require('path'),
-    utils = require('../utils');
+    utils = require('../utils'),
+    mkdirp = require('mkdirp');
 
-var generate = function(toGenerate, path) {
-  var projectRoot = utils.findProjectRoot(path);
+var generate = function(type, name) {
+  var projectRoot = utils.findProjectRoot(process.cwd());
   if (projectRoot == '') {
     process.stderr.write('Could not find nin.json in project root\n');
     process.exit(1);
   }
 
-  if (endsWith(toGenerate, 'Layer')) {
-    generateLayer(toGenerate, projectRoot);
-    addToLayers(toGenerate, projectRoot);
+  if (name == "") {
+    return;
+  }
+
+  var camelizedName = utils.camelize(name);
+
+  switch (type) {
+    case 'layer':
+      generateLayer(camelizedName, 'TemplateLayer.js',
+          [[/TemplateLayer/g, camelizedName]],
+          projectRoot);
+
+      addToLayers(camelizedName, 'red', projectRoot);
+      break;
+
+    case 'shader':
+      generateShader(camelizedName, projectRoot);
+      break;
+
+    case 'shaderWithLayer':
+      var shaderLayerName = camelizedName + 'Layer';
+
+      generateShader(camelizedName, projectRoot);
+      generateLayer(shaderLayerName, 'TemplateShaderLayer.js',
+          [[/TemplateLayer/g, shaderLayerName],
+           [/TemplateShader/g, camelizedName]],
+          projectRoot);
+
+      addToLayers(camelizedName + 'Layer', 'pink', projectRoot);
+      break;
+
+    default:
+      process.stderr.write('Attempted to generate resource without generator:', type, '\n');
   }
 }
 
-var generateLayer = function(layerName, projectRoot) {
+var generateShader = function(shaderName, projectRoot) {
+  var targetShaderPath = path.join(projectRoot, 'src', 'shaders', shaderName),
+      templateShaderPath = path.join(__dirname, 'templateShader');
+
+  mkdirp.sync(targetShaderPath);
+  fs.readdirSync(templateShaderPath).forEach(function (fileName) {
+    var from = path.join(templateShaderPath, fileName),
+        to = path.join(targetShaderPath, fileName);
+    fs.createReadStream(from).pipe(fs.createWriteStream(to));
+  });
+
+  process.stdout.write('Generated shader ' + shaderName + '\n');
```

Comment:
this puts it in stdout, while the other (stderr) in stderr.
More correct for console programs i feel.